### PR TITLE
activation threshold tutorial 

### PR DESCRIPTION
**Pull request by samuel-wj-chapman**
## Pull Request Description:
Revised activation threshold tutorial based on feedback.

commits include tutorial tools to assist with future tutorial dev, activation threshold walk through.
tutorial guides user through the process of feeding a representative dataset through a model and gaining outputs from specifc layers, plotting distributions along with mct's respective thresholds. measuring accuracy of two methods of error calculation

## Checklist before requesting a review:
- [ x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [ x] All function and files are well documented. 
- [ x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [ x] The function and variable names are informative. 
- [ x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+      "outputs": [],
+      "source": [
+        "from model_compression_toolkit import QuantizationErrorMethod\n",
+        "\n",
+        "# Specify the IMX500-v1 target platform capability (TPC)\n",
+        "tpc = mct.get_target_platform_capabilities(\"tensorflow\", 'imx500', target_platform_version='v1')\n",
+        "\n",
+        "# Set the following quantization configurations:\n",
+        "# Choose the desired QuantizationErrorMethod for the quantization parameters search.\n",
+        "# Enable weights bias correction induced by quantization.\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:367]: why do we need z threshold and shift negative in this tutorial? I think it is just confusing because its not mandatory for threshold selection)
Just use the default QuantizationConfig and override only the error method (because this is what we are trying to demonstrate in this tutorial)

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+      ],
+      "metadata": {
+        "id": "A8UHRsh2khM4"
+      },
+      "id": "A8UHRsh2khM4"
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "To assist with understanding we will now plot the threshold and distributuion from two of Mobilenetv2's layers.\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:477]: This paragraph needs to include more explanation regarding what we're going to see here, how it is affected by the threshold selection method and why it is helpful.

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+    {
+      "cell_type": "markdown",
+      "id": "c38d28f3-c947-4c7c-aafa-e96cc3864277",
+      "metadata": {
+        "id": "c38d28f3-c947-4c7c-aafa-e96cc3864277"
+      },
+      "source": [
+        "First activation layer in model is 'Conv1_relu'.\n",
+        "\n",
+        "For this particular model expanded_conv_project_BN demonstrates the difference between the two error metrics so we will also use this.\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:513]: How is this layer demonstrates it? why this layer? this part need more background because this name (expanded_conv_project_BN) doesn't say anything to someone who is not an expert on MobileNetV2 architecture (and even then, because it is kind of an inner name that we give to the layer when building the model in MCT).

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb (hunk 3 above)
Comment (samuel-wj-chapman) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: the layer was found through testing iv now pointed this out in the line. i have also added a couple of words on the significance of this particular layer though, Im unsure if its worth keeping in as i think it just adds confusion.

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+    {
+      "cell_type": "markdown",
+      "id": "01c1645e-205c-4d9a-8af3-e497b3addec1",
+      "metadata": {
+        "id": "01c1645e-205c-4d9a-8af3-e497b3addec1"
+      },
+      "source": [
+        "\n",
+        "\n",
+        "Copyright 2022 Sony Semiconductor Israel, Inc. All rights reserved.\n",
```
[the remark this answers is not in the document]
Comment (samuel-wj-chapman) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: iv updated to 2024. will double check with idan monday.

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 5]
```diff
@@ -0,0 +1,866 @@
...
+        "\n",
+        "$Q(X, t, n_b)$: The quantized approximation of X, given a threshold t and bit width nb.\n",
+        "\n",
+        "$X$: The original activation tensor before quantization.\n",
+        "\n",
+        "$t$: The quantization threshold, a critical parameter for controlling the quantization process.\n",
+        "\n",
+        "$n_b$: The number of bits used in the quantization process, affecting the model's precision and size.\n",
+        "\n",
+        "To increase efficiency in calculating the threshold, the search space for best threshold is restricted to **Power of Two** values only. This both restricts number of potential values to a reasonable number and increases hardware efficency.\n",
```
[the remark this answers is not in the document]
Comment (samuel-wj-chapman) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: idan provided to changes for this section that i have just fed directly in

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 6]
```diff
@@ -0,0 +1,866 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "9c0e9543-d356-412f-acf1-c2ecad553e06",
+      "metadata": {
+        "id": "9c0e9543-d356-412f-acf1-c2ecad553e06"
+      },
+      "source": [
+        "Error methods supported by MCT;\n",
```
[the remark this answers is not in the document]
Comment (samuel-wj-chapman) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: i think listing the methods makes more sense joined to the equation section. will leave as is for now unless you feel strongly about it

**Review by samuel-wj-chapman — COMMENTED** [review]
Thankyou for the feedback. i have visited all advised changes and either updated or commented.

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 7]
```diff
@@ -0,0 +1,866 @@
+{
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "id": "f8194007-6ea7-4e00-8931-a37ca2d0dd20",
+      "metadata": {
+        "id": "f8194007-6ea7-4e00-8931-a37ca2d0dd20"
+      },
+      "source": [
+        "# Activation Threshold Demonstration For Post-Training Quantization\n",
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: Yes I think example_keras_activation_threshold_search is better

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: Listing the methods is great, I just meant that you give it a proper highlight because it's getting lost in the long introduction to the section.
Maybe just make the sentence: "Error methods supported by MCT" the title of a subsection

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb (hunk 3 above)
Comment (ofirgo) [reply tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb]: I think the addition is good and helpful for the understanding of what you're demonstrating. I would advise to keep it.

File: tutorials/resources/utils/keras_tutorial_tools.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (ofirgo) [inline tutorials/resources/utils/keras_tutorial_tools.py:1]: change year to 2024

File: tutorials/resources/utils/keras_tutorial_tools.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    # Preprocess the dataset
+    dataset = dataset.map(lambda x, y: (imagenet_preprocess_input(x, y)))
+
+    # Take the calculated number of samples (adjusted for batch size)
+    dataset = dataset.take(samples_to_take // batch + (1 if samples_to_take % batch else 0))
+
+    return dataset
+
+
+def get_representative_dataset(fraction, REPRESENTATIVE_DATASET_FOLDER, batch) -> Generator:
```
Comment (ofirgo) [inline tutorials/resources/utils/keras_tutorial_tools.py:70]: another argument in capital letters :)

**Review by ofirgo — CHANGES_REQUESTED** [review]
Several additional minor comments to fix/think about

File: tutorials/resources/utils/keras_tutorial_tools.py (hunk 9 above)
Comment (samuel-wj-chapman) [reply tutorials/resources/utils/keras_tutorial_tools.py:70]: ah :/

**Review by Idan-BenAmi — APPROVED** [review]

File: tutorials/resources/utils/keras_tutorial_tools.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+def imagenet_preprocess_input(images, labels):
+    return tf.keras.applications.mobilenet_v2.preprocess_input(images), labels
+
+def get_validation_dataset_fraction(fraction, test_dataset_folder, batch) -> tf.data.Dataset:
+    """
+    Load a fraction of the validation dataset for evaluation.
+
+    Args:
+        fraction (float, optional): Fraction of the dataset to load. Defaults to 1.0 (i.e., the entire dataset).
```
Comment (ofirgo) [inline tutorials/resources/utils/keras_tutorial_tools.py:34]: missing  test_dataset_folder, batch in args of function description

File: tutorials/resources/utils/keras_tutorial_tools.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+    return dataset
+
+
+def get_representative_dataset(fraction, representative_dataset_folder, batch) -> Generator:
+    """
+    A function that loads a fraction of the dataset and returns a representative dataset generator.
+
+    Args:
+        fraction (float): The fraction of the dataset to load. Defaults to 1.0 (the entire dataset).
```
Comment (ofirgo) [inline tutorials/resources/utils/keras_tutorial_tools.py:76]: same

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold_search.ipynb [hunk 12]
```diff
@@ -0,0 +1,835 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "id": "9c0e9543-d356-412f-acf1-c2ecad553e06",
+      "metadata": {
+        "id": "9c0e9543-d356-412f-acf1-c2ecad553e06"
+      },
+      "source": [
+        "## Error methods supported by MCT:\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold_search.ipynb:103]: can it be a subsection? (the title is a bit big)

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 13]
```diff
@@ -0,0 +1,848 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "NGnjrPD_uTd5",
+      "metadata": {
+        "id": "NGnjrPD_uTd5"
+      },
+      "outputs": [],
+      "source": [
+        "optimal_thresholds_relu = {error_method: data[\"quantized_model\"].layers[4].activation_holder_quantizer.get_config()['threshold'][0] for error_method, data in quantized_models_dict.items()}\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:572]: break this lines to several parts so it would be possible to see all the text

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb [hunk 14]
```diff
@@ -0,0 +1,848 @@
...
+      "execution_count": null,
+      "id": "VPb8tBNGpJjo",
+      "metadata": {
+        "id": "VPb8tBNGpJjo"
+      },
+      "outputs": [],
+      "source": [
+        "import matplotlib.pyplot as plt\n",
+        "import numpy as np\n",
+        "optimal_thresholds_relu = {error_method: data[\"quantized_model\"].layers[4].activation_holder_quantizer.get_config()['threshold'][0] for error_method, data in quantized_models_dict.items()}\n",
```
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:601]: same, break the line

File: tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb (hunk 13 above)
Comment (ofirgo) [inline tutorials/notebooks/keras/ptq/example_keras_activation_threshold.ipynb:572]: please add a comment indicating which layers are 4 and 9

**Review by ofirgo — APPROVED** [review]
Additional minor comments.
I'm approving the PR but please address these last comments.
